src/js/gi/containers/GiBillApp.jsx

```diff
@@ -16,18 +17,60 @@ const Disclaimer = () => {
 };
 
 class GiBillApp extends React.Component {
+  constructor(props) {
+    super(props);
+    this.exitPreviewMode = this.exitPreviewMode.bind(this);
+  }
+
+  componentDidMount() {
+    this.props.fetchConstants(this.props.location.query.version);
+  }
+
+  componentDidUpdate(prevProps) {
+    const {
+      preview,
+      location: { query: { version: uuid } }
+    } = this.props;
+
+    const shouldHidePreviewMode = preview.display && !uuid;
+    const shouldShowPreviewMode = !preview.display && uuid && preview.version.createdAt;
+
+    if (shouldHidePreviewMode) {
+      this.props.exitPreviewMode();
+    } else if (shouldShowPreviewMode) {
+      this.props.enterPreviewMode();
+    }
+
+    const shouldRefetchConstants = prevProps.location.query.version !== uuid;
 
-  componentWillMount() {
-    this.props.updateConstants();
+    if (shouldRefetchConstants) {
+      this.props.fetchConstants(uuid);
```

Comment:
Constants don't really have versioning right now. The current purpose of this is to reset the `preview.version` object in state, in case the user goes back to preview mode right after exiting (going back in history so the query param gets tacked on again). But if versioning gets introduced later, there won't need to be a change here, so that's a bonus.